calico.go

```diff
@@ -319,24 +319,40 @@ func cmdDel(args *skel.CmdArgs) error {
 		return err
 	}
 
-	if err := calicoClient.WorkloadEndpoints().Delete(api.WorkloadEndpointMetadata{
+	ep := api.WorkloadEndpointMetadata{
 		Name:         args.IfName,
 		Node:         hostname,
 		Orchestrator: orchestrator,
-		Workload:     workload}); err != nil {
-		return err
+		Workload:     workload}
+	if err = calicoClient.WorkloadEndpoints().Delete(ep); err != nil {
+		if _, ok := err.(errors.ErrorResourceDoesNotExist); ok {
+			logger.WithField("endpoint", ep).Warn("Asked to delete endpoint object, but it doesn't exist. Ignoring.")
+		} else {
+			return err
+		}
 	}
 
 	// Only try to delete the device if a namespace was passed in.
 	if args.Netns != "" {
-		fmt.Fprintf(os.Stderr, "Calico CNI deleting device in netns %s\n", args.Netns)
-		err = ns.WithNetNSPath(args.Netns, func(_ ns.NetNS) error {
-			_, err = ip.DelLinkByNameAddr(args.IfName, netlink.FAMILY_V4)
+		logger.Info("Checking namespace & device exist.")
+		devErr := ns.WithNetNSPath(args.Netns, func(_ ns.NetNS) error {
+			_, err := netlink.LinkByName(args.IfName)
 			return err
 		})
 
-		if err != nil {
-			return err
+		if devErr == nil {
+			fmt.Fprintf(os.Stderr, "Calico CNI deleting device in netns %s\n", args.Netns)
+			err = ns.WithNetNSPath(args.Netns, func(_ ns.NetNS) error {
+				_, err = ip.DelLinkByNameAddr(args.IfName, netlink.FAMILY_V4)
+				return err
+			})
+
+			if err != nil {
+				return err
+			}
+		} else {
+			// This commonly occurs if Calico was issued a teardown but never received the SetUp signal.
+			logger.Warn("Asked to clean up container namespace, but container device (or namespace) does not exist. Ignoring.")
```

Comment:
something more benign sounding would be nice too: "veth does not exist, no need to clean up"